### ISSUE TITLE
Cache key to include Match length

### ISSUE THREAD

**Issue by drvinceknight** [Axelrod-Python/Axelrod#533]
Following #527 we have a nice cache class. This currently is designed for tournaments of constant match length and maps: pairs of players (when both players are not stochastic) to outcomes.

```
(axelrod.Cooperator, axelrod.Alternator): [('C', 'C'), ('C', 'D'), ('C', 'C')]
(axelrod.TitForTat, axelrod.Alternator): [('C', 'C'), ('C', 'D'), ('D', 'C')]
```

A small change would allow us to map pairs of players and match lengths to outcomes:

```
(axelrod.Cooperator, axelrod.Alternator, 3): [('C', 'C'), ('C', 'D'), ('C', 'C')]
(axelrod.TitForTat, axelrod.Alternator, 5): [('C', 'C'), ('C', 'D'), ('D', 'C'), ('C', 'D'), ('D', 'C')]
```

The main goal of this would be to continue the move away from assumptions about the tournament structure which will continue to make new tournaments easily. For example it would be easy to create a tournament where half matches are a single turn and the other half are a full 200 (or whatever).

Basically, whenever a match between deterministic players of any length happens, the cache could be used (or updated). The structure of the tournament would be irrelevant.
